### PR TITLE
Graphql mutations ja

### PR DESCRIPTION
**Pull request by JavkhlanShirendev**
`sendNotifications` method on channels is not yet implemented.
Please review the latest changes.
There is some weird occasional 1, 2 test fails every 2nd or 3rd time i run the tests related to async and await inside model static methods. (for example `Form.duplicate`)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,6 +33,7 @@
     "cors": "^2.8.1",
     "dotenv": "^4.0.0",
     "express": "^4.15.2",
+    "faker": "^4.1.0",
```
Comment (batamar) [inline package.json]: Remove this line, We have this package in devDependencies section

File: package.json [hunk 2]
```diff
@@ -57,6 +59,7 @@
     "jest": "^21.2.1",
     "lint-staged": "^3.6.0",
     "nodemon": "^1.11.0",
-    "prettier": "^1.4.4"
+    "prettier": "^1.4.4",
```
Comment (batamar) [inline package.json]: Remove this line, We have this package in devDependencies section

File: package.json [hunk 3]
```diff
@@ -57,6 +59,7 @@
...
+    "shortid": "^2.2.8"
```
Comment (batamar) [inline package.json]: Remove this line, We are using meteor-random temporarily

File: src/__tests__/formMutations.test.js [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+  /**
+   * Deleting the data that was used in test
+   */
+  afterEach(async () => {
+    await Users.remove({});
+    await Forms.remove({});
+  });
+
+  test('form removal test', async () => {
```
Comment (batamar) [inline src/__tests__/formMutations.test.js]: When someone try to remove form with some fields, It must raise error. Check this situation

File: src/data/schema/brand.js [hunk 5]
```diff
@@ -10,6 +10,13 @@ export const types = `
   }
```
Comment (batamar) [inline src/data/schema/brand.js]: Remove this file and get latest changes from feature-graphql-mutations branch. We have brand schema already

File: src/data/schema/channel.js [hunk 6]
```diff
@@ -16,3 +16,22 @@ export const queries = `
...
+
+export const mutations = `
+  channelsCreate(
+    name: String!,
+    description: String,
+    memberIds: [String],
+    integrationIds: [String],
+    userId: String!): Channel
+
+  channelsUpdate(
```
Comment (batamar) [inline src/data/schema/channel.js]: rename to channelsEdit

File: src/db/models/Channels.js [hunk 7]
```diff
@@ -1,22 +1,88 @@
 import mongoose from 'mongoose';
-import Random from 'meteor-random';
+import shortid from 'shortid';
```
Comment (batamar) [inline src/db/models/Channels.js]: use meteor random

File: src/db/models/Forms.js [hunk 8]
```diff
@@ -1,35 +1,146 @@
...
+
+  static updateForm(_id, doc) {
+    return this.update({ _id }, doc);
+  }
+
+  static async removeForm(_id) {
+    const fieldCount = await FormFields.find({ formId: _id }).count();
+
+    if (fieldCount > 0) {
+      throw 'You cannot delete this form. This form has some fields.';
```
Comment (batamar) [inline src/db/models/Forms.js]: throw Error ?

File: src/db/models/Forms.js [hunk 9]
```diff
@@ -1,35 +1,146 @@
...
+    }
+
+    const integrationCount = await Integrations.find({ formId: _id }).count();
+
+    if (integrationCount > 0) {
+      throw 'You cannot delete this form. This form used in integration.';
```
Comment (batamar) [inline src/db/models/Forms.js]: throw Error ?

File: src/db/models/Forms.js [hunk 10]
```diff
@@ -1,35 +1,146 @@
...
 
-export const FormFields = mongoose.model('form_fields', FieldSchema);
+class FormField {
+  static async createFormField(formId, doc) {
+    const lastField = await FormFields.findOne({}, { order: 1 }, { sort: { order: -1 } });
+    doc.formId = formId;
+
+    // if there is no field then start with 0
+    let order = 0;
+    if (lastField) {
```
Comment (batamar) [inline src/db/models/Forms.js]: New line above and below

File: src/db/models/Integrations.js [hunk 11]
```diff
@@ -1,23 +1,182 @@
...
-    default: () => Random.id(),
+    default: shortid.generate,
   },
   kind: String,
   name: String,
   brandId: String,
   formId: String,
-  formData: Object,
-  messengerData: Object,
+  formData: FormDataSchema,
```
Comment (batamar) [inline src/db/models/Integrations.js]: Nice refactor

File: src/db/models/Integrations.js [hunk 12]
```diff
@@ -1,23 +1,182 @@
...
+    return this.create(doc);
+  }
+
+  static updateFormIntegration(id, { formData, ...mainDoc }) {
+    const doc = this.generateFormDoc(mainDoc, formData);
+    return this.update({ _id: id }, doc, { runValidators: true });
+  }
+
+  static async removeIntegration(id) {
+    // remove messages
```
Comment (batamar) [inline src/db/models/Integrations.js]: What does this comment mean ?

File: src/db/models/Integrations.js [hunk 13]
```diff
@@ -1,23 +1,181 @@
...
+      kind: KIND_CHOICES.FORM,
+      formData,
+    };
+  }
+
+  static createIntegration(doc) {
+    return this.create(doc);
+  }
+
+  static createMessengerIntegration({ name, brandId }) {
```
Comment (batamar) [inline src/db/models/Integrations.js]: use doc instead of { name, brandId }

File: src/db/models/Integrations.js [hunk 14]
```diff
@@ -1,23 +1,181 @@
...
+    return this.createIntegration({
+      name,
```
Comment (batamar) [inline src/db/models/Integrations.js]: ...doc,
kind: KIND_CHOICES.MESSENGER

File: src/db/models/Integrations.js [hunk 15]
```diff
@@ -1,23 +1,181 @@
...
+      brandId,
+      kind: KIND_CHOICES.MESSENGER,
+    });
+  }
+
+  static updateMessengerIntegration(id, { name, brandId }) {
```
Comment (batamar) [inline src/db/models/Integrations.js]: _id, doc
this.update({ _id }, { $set: doc }, { runValidators: true });

File: src/db/models/Integrations.js [hunk 16]
```diff
@@ -1,23 +1,181 @@
...
+    return this.update({ _id: id }, { name, brandId }, { runValidators: true });
+  }
+
+  static saveMessengerAppearanceData(id, { color, wallpaper, logo }) {
```
Comment (batamar) [inline src/db/models/Integrations.js]: use doc

File: src/__tests__/integrationMutations.test.js [hunk 17]
```diff
@@ -0,0 +1,532 @@
...
+  let _form;
+  let _form2;
+
+  beforeEach(async () => {
+    _brand = await brandFactory({});
+    _brand2 = await brandFactory({});
+    _user = await userFactory({});
+    _form = await formFactory({ createdUserId: _user._id });
+    _form2 = await formFactory({ createdUserId: _user._id });
+  }),
```
Comment (batamar) [inline src/__tests__/integrationMutations.test.js]: What is this , ?

File: src/__tests__/integrationMutations.test.js [hunk 18]
```diff
@@ -0,0 +1,532 @@
...
+    _brand = await brandFactory({});
+    _integration = await integrationFactory({
+      name: 'messenger integration test',
+      brandId: _brand._id,
+      kind: 'messenger',
+    });
+  });
+
+  /**
+   */
```
Comment (batamar) [inline src/__tests__/integrationMutations.test.js]: Empty ?

File: src/__tests__/integrationMutations.test.js [hunk 19]
```diff
@@ -0,0 +1,532 @@
...
+    expect(integration.uiOptions.logo).toEqual(uiOptions.logo);
+  });
+});
+
+describe('save integration messenger configurations test', () => {
+  let _brand;
+  let _integration;
+
+  /**
+   */
```
Comment (batamar) [inline src/__tests__/integrationMutations.test.js]: Empty ?

File: src/__tests__/integrationMutations.test.js (hunk 17 above)
Comment (JavkhlanShirendev) [reply src/__tests__/integrationMutations.test.js]: oops

File: src/db/models/Integrations.js [hunk 20]
```diff
@@ -1,23 +1,181 @@
...
+    const doc = this.generateFormDoc(mainDoc, formData);
+
+    if (Object.keys(formData || {}).length === 0) {
+      throw 'formData must be supplied';
+    }
+
+    return this.create(doc);
+  }
+
+  static updateFormIntegration(id, { formData, ...mainDoc }) {
```
Comment (batamar) [inline src/db/models/Integrations.js]: rename to _id

File: src/__tests__/formMutations.test.js [hunk 21]
```diff
@@ -0,0 +1,580 @@
...
+   */
+  afterEach(async () => {
+    await Users.remove({});
+    await Forms.remove({});
+    await FormFields.remove({});
+  });
+
+  test('try to remove form with fields in it', async () => {
+    expect.assertions(2);
+    const form = await Forms.createForm({
```
Comment (batamar) [inline src/__tests__/formMutations.test.js]: Add new line below

File: src/__tests__/formMutations.test.js [hunk 22]
```diff
@@ -0,0 +1,580 @@
...
+  /**
+   * Deleting the data that was used in test
+   */
+  afterEach(async () => {
+    await Users.remove({});
+    await FormFields.remove({});
+    await Forms.remove({});
+  });
+
+  test('test of update order of form fields', async () => {
```
Comment (batamar) [inline src/__tests__/formMutations.test.js]: Remove duplicated checks

File: src/__tests__/notificationMutations.test.js [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+    // Another notification configuration
+    doc.notifType = 'conversationAssigneeChange';
+
+    notificationConfigurations = await NotificationConfigurations.createOrUpdateConfiguration(doc);
+
+    expect(notificationConfigurations.notifType).toEqual(doc.notifType);
+    expect(notificationConfigurations.isAllowed).toEqual(doc.isAllowed);
+    expect(notificationConfigurations.user).toEqual(doc.user);
+
+    // Change notification
```
Comment (batamar) [inline src/__tests__/notificationMutations.test.js]: ?

File: src/__tests__/notificationMutations.test.js [hunk 24]
```diff
@@ -0,0 +1,280 @@
...
+
+describe('test mutations', () => {
+  beforeEach(() => {});
+
+  afterEach(async () => {
+    await Notifications.remove({});
+    await NotificationConfigurations.remove({});
+  });
+
+  test('mutations', async () => {
```
Comment (batamar) [inline src/__tests__/notificationMutations.test.js]: Remove duplicated checks

File: src/db/models/Notifications.js [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+
+    // If already inserted then raise error
+    if (oldOne) {
+      await this.update({ _id: oldOne._id }, { $set: { isAllowed } });
+      return await this.findOne({ _id: oldOne._id });
+    }
+
+    // If it is first time then insert
+    selector.isAllowed = isAllowed;
+    return await this.create(selector);
```
Comment (batamar) [inline src/db/models/Notifications.js]: New line above

File: src/db/models/Forms.js [hunk 26]
```diff
@@ -1,35 +1,148 @@
...
   },
-  formId: String,
   type: String,
   validation: String,
   text: String,
   description: String,
   options: [String],
   isRequired: Boolean,
+  formId: {
+    required: true,
```
Comment (batamar) [inline src/db/models/Forms.js]: Remove required

File: src/db/models/Channels.js [hunk 27]
```diff
@@ -1,22 +1,85 @@
...
 const ChannelSchema = mongoose.Schema({
   _id: {
     type: String,
-    unique: true,
     default: () => Random.id(),
   },
-  name: String,
+  name: {
+    type: String,
+    required: true,
```
Comment (batamar) [inline src/db/models/Channels.js]: Remove required

File: src/__tests__/channelMutations.test.js [hunk 28]
```diff
@@ -0,0 +1,247 @@
...
+    expect(channel.memberIds[0]).toBe(_user2._id);
+    expect(channel.memberIds[1]).toBe(_user._id);
+    expect(channel.integrationIds.length).toEqual(1);
+    expect(channel.integrationIds[0]).toEqual(_integration._id);
+    expect(channel.userId).toEqual(doc.userId);
+    expect(channel.conversationCount).toEqual(0);
+    expect(channel.openConversationCount).toEqual(0);
+
+    doc.memberIds = [_user._id];
+    await Channels.updateChannel(channel._id, doc);
```
Comment (batamar) [inline src/__tests__/channelMutations.test.js]: Spaces ?

File: src/__tests__/integrationMutations.test.js [hunk 29]
```diff
@@ -0,0 +1,557 @@
...
+
+    expect(integrationCount).toEqual(0);
+  });
+});
+
+describe('save integration messenger appearance test', () => {
+  let _brand;
+  let _integration;
+  /**
+   */
```
Comment (batamar) [inline src/__tests__/integrationMutations.test.js]: Empty ?

File: src/db/models/Integrations.js [hunk 30]
```diff
@@ -1,23 +1,315 @@
...
+  static updateFormIntegration(_id, { formData, ...mainDoc }) {
+    const doc = this.generateFormDoc(mainDoc, formData);
+
+    return this.update({ _id }, { $set: doc }, { runValidators: true });
+  }
+
+  /**
+   * Remove integration in addition with its messages, conversations, customers
+   * @param {string} id - Integration id
+   * @return {Promise} returns null promise
```
Comment (batamar) [inline src/db/models/Integrations.js]: Not null promise

File: src/db/models/Integrations.js [hunk 31]
```diff
@@ -1,23 +1,315 @@
...
+   * @param {string} args.mainDoc - Integration main document object
+   * @param {string} args.mainDoc.name - Integration name
+   * @param {string} args.mainDoc.brandId - Integration brand id
+   * @param {string} args.mainDoc.formId - Form id related to this integration
+   * @return {Promise} returns null promise
+   */
+  static updateFormIntegration(_id, { formData, ...mainDoc }) {
+    const doc = this.generateFormDoc(mainDoc, formData);
+
+    return this.update({ _id }, { $set: doc }, { runValidators: true });
```
Comment (batamar) [inline src/db/models/Integrations.js]: return updated integration
await this.update ....
return this.findOne({ _id })

File: src/db/models/Integrations.js [hunk 32]
```diff
@@ -1,23 +1,315 @@
...
+   * @param {Email[]} doc.formData.adminEmails - TODO: need more elaborate documentation
+   * @param {string} doc.formData.adminEmailTitle - TODO: need more elaborate documentation
+   * @param {string} doc.formData.adminEmailContent - TODO: need more elaborate documentation
+   * @param {string} doc.formData.thankContent - TODO: need more elaborate documentation
+   * @param {string} doc.formData.redirectUrl - Form redirectUrl on submit
+   * @param {string} args.mainDoc - Integration main document object
+   * @param {string} args.mainDoc.name - Integration name
+   * @param {string} args.mainDoc.brandId - Integration brand id
+   * @param {string} args.mainDoc.formId - Form id related to this integration
+   * @return {Promise} returns null promise
```
Comment (batamar) [inline src/db/models/Integrations.js]: updated integration

File: src/db/models/Integrations.js [hunk 33]
```diff
@@ -1,23 +1,315 @@
...
+  *    TODO: need more elaborate documentation
+  * @param {string} doc.messengerData.thankYouMessage - Thank you message
+  *    TODO: need more elaborate documentation
+  * @param {string} doc.messengerData.uiOptions.color - Color of messenger
+  * @param {string} doc.messengerData.uiOptions.wallpaper - Wallpaper image for messenger
+  * @param {string} doc.messengerData.uiOptions.logo - Logo used in the embedded messenger
+  * @return {Promise} returns null promise
+  */
+  static saveMessengerConfigs(_id, messengerData) {
+    return this.update({ _id }, { $set: { messengerData } }, { runValidators: true });
```
Comment (batamar) [inline src/db/models/Integrations.js]: return updated integration
await this.update ....
return this.findOne({ _id })

File: src/db/models/Integrations.js [hunk 34]
```diff
@@ -1,23 +1,315 @@
...
+   * Save messenger appearance data
+   * @param {string} _id
+   * @param {Object} object - MessengerUiOptions object TODO: need more elaborate documentation
+   * @param {string} object.color - MessengerUiOptions color TODO: need more elaborate documentation
+   * @param {string} object.wallpaper - MessengerUiOptions wallpaper
+   * @param {string} object.logo - Messenger logo TODO: need more elaborate documentation
+   * @return {Promise} returns null promise
+   */
+  static saveMessengerAppearanceData(_id, { color, wallpaper, logo }) {
+    return this.update(
```
Comment (batamar) [inline src/db/models/Integrations.js]: return updated integration
await this.update ....
return this.findOne({ _id })

File: src/db/models/Integrations.js [hunk 35]
```diff
@@ -1,23 +1,315 @@
...
+
+  /**
+   * Update messenger integration document
+   * @param {Object} object - Integration main doc object
+   * @param {string} object.name - Integration name
+   * @param {string} object.brandId - Integration brand id
+   * @return {Promise} return null promise
+   */
+  static updateMessengerIntegration(_id, { name, brandId }) {
+    return this.update({ _id }, { $set: { name, brandId } }, { runValidators: true });
```
Comment (batamar) [inline src/db/models/Integrations.js]: return updated integration
await this.update ....
return this.findOne({ _id })

File: src/db/models/Channels.js [hunk 36]
```diff
@@ -1,22 +1,105 @@
...
+  static updateChannel(_id, doc) {
+    const { userId } = doc;
+
+    if (userId && userId._id) {
+      doc.userId = doc.userId._id;
+    }
+
+    this.preSave(doc);
+
+    return this.update({ _id }, { $set: doc }, { runValidators: true });
```
Comment (batamar) [inline src/db/models/Channels.js]: return updated integration
await this.update ....
return this.findOne({ _id })

File: src/db/models/Forms.js [hunk 37]
```diff
@@ -1,35 +1,235 @@
...
+  /**
+   * Updates a form document
+   * @param {string} _id - Form id
+   * @param {Object} object - Form object
+   * @param {string} object.title - Form title
+   * @param {string} object.description - Form description
+   * @return {Promise} returns null
+   */
+  static updateForm(_id, { title, description }) {
+    return this.update({ _id }, { $set: { title, description } }, { runValidators: true });
```
Comment (batamar) [inline src/db/models/Forms.js]: return updated form
await this.update ....
return this.findOne({ _id })

File: src/db/models/Forms.js [hunk 38]
```diff
@@ -1,35 +1,235 @@
...
+    return this.create(doc);
+  }
+
+  /**
+   * Updates a form document
+   * @param {string} _id - Form id
+   * @param {Object} object - Form object
+   * @param {string} object.title - Form title
+   * @param {string} object.description - Form description
+   * @return {Promise} returns null
```
Comment (batamar) [inline src/db/models/Forms.js]: updated form not null promise

File: src/db/models/Notifications.js [hunk 39]
```diff
@@ -0,0 +1,154 @@
...
+   * @param {string} doc.notifType - Category of notification  (module)
+   * @param {string} doc.createdUser - The user whose actions made this notification
+   * @param {string} doc.title - Notificaton title
+   * @param {string} doc.content - Notification content
+   * @param {string} doc.link - Notification link
+   * @param {Object|string} doc.receiver - Id of the user that will receive this notification
+   * @return {Promise} The promise returns null
+   */
+  static updateNotification(_id, doc) {
+    return this.update({ _id }, doc);
```
Comment (batamar) [inline src/db/models/Notifications.js]: return updated notification
await this.update ....
return this.findOne({ _id })

File: src/db/models/Notifications.js [hunk 40]
```diff
@@ -0,0 +1,154 @@
...
+   * Update a notification
+   * @param {string} _id - Id of notification
+   * @param {Object} doc - Notification object
+   * @param {string} doc.notifType - Category of notification  (module)
+   * @param {string} doc.createdUser - The user whose actions made this notification
+   * @param {string} doc.title - Notificaton title
+   * @param {string} doc.content - Notification content
+   * @param {string} doc.link - Notification link
+   * @param {Object|string} doc.receiver - Id of the user that will receive this notification
+   * @return {Promise} The promise returns null
```
Comment (batamar) [inline src/db/models/Notifications.js]: not pull promise

File: src/db/models/Forms.js [hunk 41]
```diff
@@ -1,35 +1,235 @@
...
+   * @param {Object} doc - Form object
+   * @param {string} doc.title - Form title
+   * @param {string} doc.description - Form description
+   * @param {Date} doc.createdDate - Form creation date
+   * @param {Object|string} createdUser - The user who is creating this form,
+   * can be both user id or user object
+   * @return {Promise} returns Form document promise
+   * @throws {Error} throws Error if createdUser is not supplied
+   */
+  static async createForm(doc, createdUser) {
```
Comment (batamar) [inline src/db/models/Forms.js]: rename to createdUserId

**Pull request merged**